akka-stream/src/main/scala/akka/stream/impl/FanIn.scala

```diff
@@ -183,13 +204,19 @@ private[akka] object FanIn {
     }
 
     def inputsAvailableFor(id: Int) = new TransferState {
-      override def isCompleted: Boolean = depleted(id) || cancelled(id)
-      override def isReady: Boolean = pending(id)
+      override def isCompleted: Boolean = {
+        val state = states(id)
+        has(state, Depleted) || has(state, Cancelled)
+      }
```

Comment:
In the currently commented case it is rather obvious where `state` is coming from, but especially the while loop in `idToDequeue` would be trivial to break by not remembering that there is a `id <=> state` correspondence that needs to be maintained manually. In `dequeue` the distance between `id` and `state` and its usage is also quite big. In contrast `hasFlag(id, Depleted)` has all the pieces in one place. If it is measurably slower then there would be a reason for this reduced readability …